mod-source-record-manager-server/src/main/java/org/folio/services/flowcontrol/RawRecordsFlowControlServiceImpl.java

```diff
@@ -0,0 +1,117 @@
+package org.folio.services.flowcontrol;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.folio.kafka.KafkaConsumerWrapper;
+import org.folio.verticle.consumers.consumerstorage.KafkaConsumersStorage;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+import javax.annotation.PostConstruct;
+import java.util.Collection;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.folio.rest.jaxrs.model.DataImportEventTypes.DI_RAW_RECORDS_CHUNK_READ;
+
+@Service
+public class RawRecordsFlowControlServiceImpl implements RawRecordsFlowControlService {
+  private static final Logger LOGGER = LogManager.getLogger();
+
+  @Value("${di.flow.control.max.simultaneous.records:100}")
+  private Integer maxSimultaneousRecords;
+  @Value("${di.flow.control.records.threshold:50}")
+  private Integer recordsThreshold;
+  @Value("${di.flow.control.enable:true}")
+  private boolean enableFlowControl;
+
+  @Autowired
+  private KafkaConsumersStorage consumersStorage;
+
+  /**
+   * Represents the current state of flow control to make decisions when need to make pause/resume calls.
+   * It is basically counter, that increasing when DI_RAW_RECORDS_CHUNK_READ event comes and decreasing when
+   * DI_COMPLETE, DI_ERROR come.
+   *
+   * When current state equals or exceeds di.flow_control.max_simultaneous_records - all raw records consumers from consumer's group should pause.
+   * When current state equals or below di.flow_control.records_threshold - all raw records consumers from consumer's group should resume.
+   */
+  private final AtomicInteger currentState = new AtomicInteger(0);
```

Comment:
General question. is handling of message done on the vertx event loop?

trying to reason to see if having this synchronization will slow things.